staging/src/k8s.io/client-go/tools/cache/controller.go

```diff
@@ -147,14 +160,26 @@ func (c *controller) Run(stopCh <-chan struct{}) {
 	c.reflector = r
 	c.reflectorMutex.Unlock()
 
-	var wg wait.Group
-
-	wg.StartWithChannel(stopCh, r.Run)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		r.RunWithStopOptions(stopOptions)
+	}()
 
-	wait.Until(c.processLoop, time.Second, stopCh)
+	// TODO: Does the process loop also need to be ran with stop options?
+	wait.Until(c.processLoop, time.Second, c.config.StopHandle.Done())
 	wg.Wait()
 }
 
+// Run supports calling RunWithStopOptions with just a stop channel
+// that when closed, is the only stop condition that will stop the controller.
```

Comment:
In the most common use case for the controller (ran by a shared informer that is started by a factory), it will never be called more than.

That being said, if the controller was being used as a library outside an informer factory, my understanding is that yes, controllers (and shared informers) can be ran more than once and that should be fine. 

I did a quick test to verify this and no issues arise, but it does create multiple reflectors that react separately to each event the reflector detects. I think this is how it’s meant to work because multiple consumers are likely to start their own copies of shared informers for a given resource.

Can you confirm @DirectXMan12

Reply:
Also, just to be clear, nothing is changing about the behavior of `Run`, any concerns about it being called multiple times are the same now as they were before if I'm understanding your question correctly.

Reply:
Oh I see. Sorry that deletion was a mistake, I added it back to Run (and RunWithStopOptions). The behavior on this should not be changed.